File: System_Design_Docs/experiments/lsm_sim.py
#!/usr/bin/env python3
"""
Minimal LSM-tree simulator — measures the three amplifications that define
RocksDB's behaviour: write, read, and space amplification.

It models the real RocksDB write path:
    writes -> in-memory MemTable -> flush to an L0 SSTable -> leveled compaction

We count actual bytes written to "disk" (every flush + every compaction rewrite),
SSTables touched per point lookup (with and without a Bloom filter), and bytes
held on disk vs live user data. No real I/O — just honest accounting of the
same operations RocksDB performs, so the amplification ratios are meaningful.
"""
import random

KEY_SPACE   = 20_000     # distinct keys (so updates overwrite -> dead versions)
N_WRITES    = 200_000    # total user writes (key,value) pairs
ENTRY_BYTES = 100        # bytes per entry (key+value+overhead)
MEMTABLE    = 2_000      # entries held in memory before a flush
LEVEL_RATIO = 10         # each level ~10x the previous (RocksDB default)
random.seed(42)

class Stats:
    def __init__(self):
        self.user_bytes = 0      # logical bytes the user wrote
        self.disk_writes = 0     # physical bytes written (flush + compaction)

def run(compaction):
    """compaction in {'none','leveled'}; returns metrics dict."""
    st = Stats()
    # levels: list of "SSTables"; each SSTable is a dict{key:val}. L0 is a list of tables.
    L0 = []
    levels = {1: {}, 2: {}, 3: {}}   # leveled: one sorted run per level (merged)
    memtable = {}

    def flush():
        # MemTable -> new L0 SSTable. Bytes written = size of the table.
        nonlocal memtable
        if not memtable:
            return
        table = dict(memtable)
        st.disk_writes += len(table) * ENTRY_BYTES
        L0.append(table)
        memtable = {}
        if compaction == 'leveled':
            compact()

    def compact():
        # Push L0 down through levels when a level exceeds its size budget.
        # Merging rewrites the merged data -> that rewrite IS write amplification.
        while len(L0) >= 4:                       # L0 trigger (RocksDB default = 4)
            merged = {}
            for t in L0:
                merged.update(t)                  # newer overwrites older
            L0.clear()
            merged.update(levels[1])
            st.disk_writes += len(merged) * ENTRY_BYTES
            levels[1] = merged
            for lvl in (1, 2):
                budget = MEMTABLE * (LEVEL_RATIO ** lvl)
                if len(levels[lvl]) > budget:
                    nxt = dict(levels[lvl + 1])
                    nxt.update(levels[lvl])
                    st.disk_writes += len(nxt) * ENTRY_BYTES
                    levels[lvl + 1] = nxt
                    levels[lvl] = {}

    # ---- write path ----
    for _ in range(N_WRITES):
        k = random.randint(1, KEY_SPACE)
        memtable[k] = _
        st.user_bytes += ENTRY_BYTES
        if len(memtable) >= MEMTABLE:
            flush()
    flush()

    # ---- count runs (for read amplification) ----
    runs = len(L0) + sum(1 for lvl in levels.values() if lvl)

    # ---- space: physical live bytes on disk vs logical live data ----
    on_disk_entries = sum(len(t) for t in L0) + sum(len(lvl) for lvl in levels.values())
    live_keys = set()
    for lvl in levels.values():
        live_keys |= lvl.keys()
    for t in L0:
        live_keys |= t.keys()
    live_entries = len(live_keys)

    # ---- read amplification: SSTables probed per lookup ----
    # Without bloom filters you may probe every run; with them you skip runs
    # that certainly don't hold the key (1% false-positive rate assumed).
    trials = 5_000
    probes_nobloom = probes_bloom = 0
    all_runs = [t for t in L0] + [lvl for lvl in levels.values() if lvl]
    for _ in range(trials):
        k = random.randint(1, KEY_SPACE)
        for run_tbl in all_runs:
            probes_nobloom += 1
            if k in run_tbl:
                probes_bloom += 1
                break
            else:
                probes_bloom += 0.01          # 1% bloom false-positive probe
    return {
        'compaction': compaction,
        'runs_to_search': runs,
        'write_amp': round(st.disk_writes / st.user_bytes, 2),
        'space_amp': round(on_disk_entries / live_entries, 2),
        'read_probes_per_lookup_nobloom': round(probes_nobloom / trials, 2),
        'read_probes_per_lookup_bloom':   round(probes_bloom / trials, 2),
    }

def show(m):
    print(f"  compaction = {m['compaction']}")
    print(f"    sorted runs to search per read : {m['runs_to_search']}")
    print(f"    WRITE amplification            : {m['write_amp']}x  (disk bytes / user bytes)")
    print(f"    SPACE amplification            : {m['space_amp']}x  (on-disk entries / live entries)")
    print(f"    READ probes/lookup (no bloom)  : {m['read_probes_per_lookup_nobloom']}")
    print(f"    READ probes/lookup (w/ bloom)  : {m['read_probes_per_lookup_bloom']}")
    print()

if __name__ == '__main__':
    print(f"LSM simulator: {N_WRITES:,} writes over {KEY_SPACE:,} keys "
          f"(={N_WRITES//KEY_SPACE}x overwrite), memtable={MEMTABLE}\n")
    print("=== No compaction (writes are cheap, reads/space blow up) ===")
    show(run('none'))
    print("=== Leveled compaction (RocksDB default — pays writes to keep reads/space low) ===")
    show(run('leveled'))

File: System_Design_Docs/RocksDB/lsm_sim.cpp
// lsm_sim.cpp — a small, faithful model of an LSM-tree storage engine
// (the architecture RocksDB uses) to measure the three amplification factors:
//   write amplification  = bytes physically written / bytes of user data
//   read amplification    = sorted runs probed per lookup (with/without bloom)
//   space amplification   = bytes on disk / bytes of live user data
//
// It is NOT RocksDB; it is a simplification that reproduces the same dynamics:
// memtable -> flush to L0 -> leveled compaction with bounded fan-out overlap,
// plus a bloom-filter model for the read path.
// Build:  c++ -O2 -std=c++17 lsm_sim.cpp -o lsm_sim
#include <cstdio>
#include <cstdint>
#include <vector>
#include <random>
#include <algorithm>
#include <functional>

int main() {
    const int  KEY_SPACE  = 200000;   // distinct keys (updates overwrite)
    const int  NUM_WRITES = 2000000;  // total put operations
    const int  VALUE_SIZE = 100;      // bytes per value
    const int  MEMTABLE   = 4000;     // entries before a flush
    const int  L0_TRIGGER = 4;        // L0 files before compaction to L1
    const int  FANOUT     = 10;       // each level ~10x previous (overlap factor)

    // Per-level entry budgets (L1 small, growing 10x). KEY_SPACE caps real data.
    long budget[5] = {0, 8000, 80000, 800000, 1L<<60};
    long lvl[5]    = {0,0,0,0,0};      // resident entries per level L1..L4

    uint64_t userBytes = 0, physBytes = 0;
    int memEntries = 0;
    std::vector<int> l0Files;
    auto B = [&](long e){ return (uint64_t)e * VALUE_SIZE; };

    // Compact `moved` entries down into level L. Real leveled compaction only
    // rewrites the overlapping slice of L (~FANOUT * moved), not all of L.
    std::function<void(int,long)> pushDown = [&](int L, long moved){
        long overlap = std::min(lvl[L], moved * FANOUT);
        physBytes += B(moved + overlap);          // rewrite merged + overlap
        lvl[L] += moved;
        if (L < 4 && lvl[L] > budget[L]) {         // over budget -> cascade
            long spill = lvl[L] - budget[L];
            lvl[L] = budget[L];
            pushDown(L+1, spill);
        }
    };

    for (int i = 0; i < NUM_WRITES; ++i) {
        userBytes += VALUE_SIZE;
        if (++memEntries >= MEMTABLE) {
            physBytes += B(memEntries);             // flush memtable -> L0 file
            l0Files.push_back(memEntries);
            memEntries = 0;
            if ((int)l0Files.size() >= L0_TRIGGER) {
                long moved = 0; for (int e: l0Files) moved += e;
                l0Files.clear();
                pushDown(1, moved);                 // compact L0 -> L1 (cascades)
            }
        }
    }

    // ---- READ PATH: sorted runs a point lookup may probe
    int sortedRuns = (memEntries>0) + (int)l0Files.size();
    for (int L=1;L<=4;++L) sortedRuns += (lvl[L]>0);

    const int    PROBES  = 200000;
    const double BLOOM_FP = 0.01;                   // 1% false-positive rate
    std::mt19937 rng(42);
    std::uniform_real_distribution<double> u(0,1);
    long touchesNoBloom = 0, touchesBloom = 0;
    for (int i=0;i<PROBES;++i){
        touchesNoBloom += sortedRuns;               // worst case: check all runs
        int touched = 1;                            // the one run that has it
        for (int r=0;r<sortedRuns-1;++r)
            if (u(rng) < BLOOM_FP) touched++;        // bloom false positive
        touchesBloom += touched;
    }

    // ---- SPACE: compaction dedups overwrites, so disk ~ settled distinct data
    long onDiskEntries = memEntries;
    for (int e: l0Files) onDiskEntries += e;
    for (int L=1;L<=4;++L) onDiskEntries += lvl[L];
    onDiskEntries = std::min(onDiskEntries, (long)KEY_SPACE * 2); // dedup cap
    uint64_t liveBytes = B(KEY_SPACE);

    printf("==== LSM-tree simulation (RocksDB-style leveled compaction) ====\n");
    printf("user puts             : %d  (value=%dB)\n", NUM_WRITES, VALUE_SIZE);
    printf("distinct keys         : %d\n", KEY_SPACE);
    printf("level residency (L1..L4 entries): %ld / %ld / %ld / %ld\n",
           lvl[1],lvl[2],lvl[3],lvl[4]);
    printf("\n-- WRITE amplification --\n");
    printf("user bytes            : %.1f MB\n", userBytes/1e6);
    printf("physical bytes written: %.1f MB\n", physBytes/1e6);
    printf("write amplification   : %.1fx\n", (double)physBytes/userBytes);
    printf("\n-- READ amplification --\n");
    printf("sorted runs on disk   : %d\n", sortedRuns);
    printf("SST touches/read no bloom : %.2f\n",(double)touchesNoBloom/PROBES);
    printf("SST touches/read w/ bloom : %.2f\n",(double)touchesBloom/PROBES);
    printf("bloom read I/O reduction  : %.0f%%\n",
           100.0*(1.0-(double)touchesBloom/touchesNoBloom));
    printf("\n-- SPACE amplification --\n");
    printf("live user data        : %.1f MB\n", liveBytes/1e6);
    printf("data on disk          : %.1f MB\n", B(onDiskEntries)/1e6);
    printf("space amplification   : %.2fx\n", (double)B(onDiskEntries)/liveBytes);
    return 0;
}
